### DISCUSSION TITLE
Difficulty with --cscript

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#60] [Q&A] [answered]
In order to explore and understand the usage of custom phase scripts, I pared down the example script on the Wiki (https://github.com/gitbls/sdm/wiki/Example:-Custom-Phase-script).  I removed everything except some basic logging with the intent of incrementally adding back in my desired customizations.  Here is my script ('.txt' added to file name as Github wouldn't accept '.sh' type). 
[sdm-Custom-Phase-script.sh.txt](https://github.com/gitbls/sdm/files/10494871/sdm-Custom-Phase-script.sh.txt)

When I ran sdm --customize with a --cscript switch the customization failed in all three phases with a message saying "/bin/bash^M: bad interpreter: No such file or directory".  Here is what I think are the relevant parts of the customization attempt (I can provide a full log if needed):
```
* Beginning sdm customization

[sudo] password for doug:             
* Mount IMG '/home/doug/Downloads/bullseye-armhf-customized.img'
mount: /dev/loop0p2 mounted on /mnt/sdm.
mount: /dev/loop0p1 mounted on /mnt/sdm/boot.
% Configure chroot to use this IMG on your ' ARM' computer
  * sdm will run more slowly due to qemu emulation
...
* Start Configuration
> Command Line: /usr/local/sdm/sdm --customize --wpa 
  /home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/wpa_supplicant.conf --keymap 
  us --locale en_US.UTF-8 --timezone America/Los_Angeles --wifi-country US --disable piwiz 
  --svc-disable userconfig --bootset serial=0,boot_behaviour=B4,boot_wait=0 --nouser --fstab 
  /home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/fstab_append.txt --extend 
  --xmb 1024 --poptions noupdate,noupgrade --cscript 
  /home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/sdm-Custom-Phase-script.sh 
  --restart /home/doug/Downloads/bullseye-armhf-customized.img
> IMG date: 2022-09-22
> IMG '/home/doug/Downloads/bullseye-armhf-customized.img' has 1380840 1K-blocks free 
> Copy sdm to /usr/local/sdm in the IMG
> Copy Custom Phase Script 
  '/home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/sdm-Custom-Phase-script.sh' 
  to /usr/local/sdm in the IMG
* Start Phase 0 image customization
*+Phase 0 sdm configuration
*+Phase 0 Network configuration
> Copy wpa supplicant config from 
  '/home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/wpa_supplicant.conf' to 
  /mnt/sdm/etc/wpa_supplicant/wpa_supplicant.conf
*+Phase 0 User configuration
*+Phase 0 System configuration
> Copy fstab extension 
  '/home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/fstab_append.txt' to 
  /mnt/sdm/etc/sdm/assets and enable for FirstBoot
*+Phase 0 Raspberry Pi-specific configuration
> Run Custom Phase Script '/mnt/sdm//usr/local/sdm/sdm-Custom-Phase-script.sh' Phase 0
/mnt/sdm//usr/local/sdm/sdm-phase0: /mnt/sdm//usr/local/sdm/sdm-Custom-Phase-script.sh: /bin/bash^M: bad interpreter: No such file or directory
* Phase 0 Completed
* Enter image '/home/doug/Downloads/bullseye-armhf-customized.img' for Phase 1
* Start Phase 1 image customization
> IMG '/home/doug/Downloads/bullseye-armhf-customized.img' has 1376536 1K-blocks free at start of Phase 1 image customization
*+Phase 1 sdm configuration
...
> Run Custom Phase Script '/usr/local/sdm/sdm-Custom-Phase-script.sh' Phase 1
/usr/local/sdm/sdm-phase1: /usr/local/sdm/sdm-Custom-Phase-script.sh: /bin/bash^M: bad interpreter: No such file or directory
...
> IMG '/home/doug/Downloads/bullseye-armhf-customized.img' has 1305652 1K-blocks free at end of Phase 1
* Phase 1 Completed
> Run Custom Phase Script '/usr/local/sdm/sdm-Custom-Phase-script.sh' post-install
/usr/local/sdm/sdm-phase1: /usr/local/sdm/sdm-Custom-Phase-script.sh: /bin/bash^M: bad interpreter: No such file or directory
> IMG '/home/doug/Downloads/bullseye-armhf-customized.img' has 1305648 1K-blocks free at end of Customization
* Enter Shell Command Prompt
root@dv6-Notebook-PC:/# exit
exit
* Customization complete
umount: /mnt/sdm/boot unmounted
umount: /mnt/sdm unmounted
```
I notice that there is a double-/ in the '> Run Custom Phase Script '/mnt/sdm//usr/local/sdm/sdm-Custom-Phase-script.sh' Phase 0' and following line which looks odd to me.  This doesn't appear in the Phase 1 and post-install steps.

Is /bin/bash not able to be found for some reason?  I checked both on the image (via sdm --explore) and the Linux Mint 21 computer on which I'm running sdm and can see /bin/bash.

Might this be somehow due to my running sdm on Linux Mint?

**Comment by gitbls** [comment, marked as answer]
Nope, really simple. "/bin/bash^M: bad interpreter: No such file or directory" tells me that you edited this file on Windows, and the file has Windows line endings on it rather than Linux line endings. Use dos2unix on Mint to convert the line endings and it will work fine.

There are many ways to fix the line endings. dos2unix on Mint will do it. If you don't like that, a quick search for: convert windows file to linux will give you 88M or so web sites telling you how to do it. 🤣 I personally use EditPad Lite on Windows, and make sure that the file is in Linux format when I save it, or more often than not, I just do all the editing on Linux.

The double slash is a buglet, but doesn't hurt anything. Bash nicely treats '//' as '/'. But will have a look at it.

**Reply by dshaw619** [reply]
Face palm!!!  I did do the original editing of that file on a Windows laptop, using VS Code.  Will have to look to see if if has a setting to save with Linux line endings.

**Reply by dshaw619** [reply]
Looks like the settings I should use in VS Code are:
https://stackoverflow.com/a/39532890
https://stackoverflow.com/a/66477980

**Comment by gitbls** [comment]
Yep, that's it. LF vs CRLF. 

By the way, I've decided for myself that I'm using a plugin instead of a custom phase script. They are effectively the same in terms of the structure, etc, but plugins have a few advantages that I like:
- Plugins can take arguments via the `--plugin` switch
- Plugins can be applied during the `--burn` command (although I don't do that for my own customizations)

Converting a custom phase script to a plugin is really simple. In fact, you can simply use a proper custom phase script as a plugin. Changes are only required if you want it to accept arguments or if you're OCD (like I am) about the logging.

That said, I have no intention of removing custom phase scripts, so you can customize your image either way.

**Reply by dshaw619** [reply]
Thanks.  I will look into plugins more.  I'll open a new topic with some questions so as to not get too off topic in this one.
